ipfs/ipfs.go

```diff
@@ -0,0 +1,316 @@
+
+const tempDir string = "/tmp/bytes.txt"
+const ipfsStorageDir string = "/ipfs-storage"
+
+func setupPlugins(externalPluginsPath string) error {
+	// Load any external plugins if available on externalPluginsPath
+	plugins, err := loader.NewPluginLoader(filepath.Join(externalPluginsPath, "plugins"))
+	if err != nil {
+		return fmt.Errorf("error loading plugins: %s", err)
+	}
+
+	// Load preloaded and external plugins
+	if err := plugins.Initialize(); err != nil {
+		return fmt.Errorf("error initializing plugins: %s", err)
+	}
+
+	if err := plugins.Inject(); err != nil {
+		return fmt.Errorf("error initializing plugins: %s", err)
+	}
+
+	return nil
+}
+
+func createTempRepo() (string, error) {
+	repoPath, err := ioutil.TempDir("", "ipfs-shell")
+	if err != nil {
+		return "", fmt.Errorf("failed to get temp dir: %s", err)
+	}
+
+	// Create a config with default options and a 2048 bit key
+	cfg, err := config.Init(ioutil.Discard, 2048)
+	if err != nil {
+		return "", err
+	}
+
+	// When creating the repository, you can define custom settings on the repository, such as enabling experimental
+	// features (See experimental-features.md) or customizing the gateway endpoint.
+	// To do such things, you should modify the variable `cfg`. For example:
+	if *FlagExp {
+		// https://github.com/ipfs/go-ipfs/blob/master/docs/experimental-features.md#ipfs-filestore
+		cfg.Experimental.FilestoreEnabled = true
+		// https://github.com/ipfs/go-ipfs/blob/master/docs/experimental-features.md#ipfs-urlstore
+		cfg.Experimental.UrlstoreEnabled = true
+		// https://github.com/ipfs/go-ipfs/blob/master/docs/experimental-features.md#ipfs-p2p
+		cfg.Experimental.Libp2pStreamMounting = true
+		// https://github.com/ipfs/go-ipfs/blob/master/docs/experimental-features.md#p2p-http-proxy
+		cfg.Experimental.P2pHttpProxy = true
+		// https://github.com/ipfs/go-ipfs/blob/master/docs/experimental-features.md#strategic-providing
+		cfg.Experimental.StrategicProviding = true
+	}
+
+	// Create the repo with the config
+	err = fsrepo.Init(repoPath, cfg)
+	if err != nil {
+		return "", fmt.Errorf("failed to init ephemeral node: %s", err)
+	}
+
+	return repoPath, nil
+}
+
+/// ------ Spawning the node
+
+// Creates an IPFS node and returns its coreAPI
+func createNode(ctx context.Context, repoPath string) (iface.CoreAPI, error) {
+	// Open the repo
+	repo, err := fsrepo.Open(repoPath)
+	if err != nil {
+		return nil, err
+	}
+
+	// Construct the node
+	nodeOptions := &core.BuildCfg{
+		Online:  true,
+		Routing: libp2p.DHTOption, // This option sets the node to be a full DHT node (both fetching and storing DHT Records)
+		// Routing: libp2p.DHTClientOption, // This option sets the node to be a client DHT node (only fetching records)
+		Permanent: true,
+		Repo:      repo,
+	}
+
+	node, err := core.NewNode(ctx, nodeOptions)
+	if err != nil {
+		return nil, err
+	}
+
+	// Attach the Core API to the constructed node
+	return coreapi.NewCoreAPI(node)
```

Comment:
instead of returning the api instance, can we use a variable and check for errors before return? that way we can shutdown on error of the api stand up

Reply:
Currently the flow is returning the API and err to the spawner functions who then return that err to the caller, this follows the idiomatic error handlng in Go. 
We can not use this structure of error handling but I would like to have a brief sync to go over this change.

Reply:
Agreed, the error will be returned from this method. Resolving

Reply:
I like the idiomatic way of retuning errors from the function.

As an aside, do you plan to convert these functions into methods on a struct in the PR?